theano/tensor/tests/test_elemwise.py

```diff
@@ -1242,6 +1242,36 @@ def func(x, y, z):
                                [numpy.asarray([-1., 0.5, 2.]), 0., 1.])
 
 
+def test_grad_useless_sum():
+    """Test absence of useless sum.
+
+    When an operation (such as T.mul) is done on a broadcastable vector and
+    a matrix, the gradient in backward path is computed for the broadcasted
+    vector. So a sum reverts the broadcasted vector to a vector. In the case
+    of operations on two vectors, the sum should not be generated.
+
+    This test checks whether there is a useless sum in the gradient
+    computations.
+    """
+    x = tensor.TensorType(theano.config.floatX, (True,))('x')
+    l = tensor.log(1.0 - tensor.nnet.sigmoid(x))[0]
+    g = tensor.grad(l, x)
+    nodes = theano.gof.graph.ops([x], [g])
+
+    f = theano.function([x], g)
```

Comment:
Our tests should pass with mode=FAST_COMPILE. I think this new test will fail in that case. Travis will confirm this. I think this extra parameter will make it work in that cases: mode=theano.compile.get_default_mode().including("canonizer")